### DISCUSSION TITLE
React Native Support

### DISCUSSION THREAD

**Discussion by zameschua** [tailwindlabs/headlessui#571] [Ideas]
I know lots of people come in here with 'Can you support X platform' so sorry for being another one of that.

React Native has become a killer platform in the past few years, and React Native Web is becoming a formidable player, allowing users use 1 codebase for cross-platform apps.

React Native support will allow us to build our design system for the entire company / project from the ground up and not need to worry about supporting components on separate platforms ever again.

Love the great work so far, keep killing it!

**Comment by RobinMalfait** [comment]
Hey! Thank you for your suggestion!
Much appreciated! 🙏

React Native will probably be even more difficult compared to other frameworks. This is because a lot of the code relies on the DOM. I'm also not familiar with React Native, but my assumption is that you don't really have a DOM so the conversion will probably be lots of work.

**Reply by corysimmons** [reply]
> This is because a lot of the code relies on the DOM.

It doesn't have to. HeadlessUI is using the ElementType type everywhere (and yes that is DOM specific), but it could be extended to include RN primitives (like `<View>` where `<div>` would render) and it might be easier than you think for HeadlessUI to support RN.

RN is in desperate need of a headless UI kit—whether it's HeadlessUI or Radix. At this point, it's literally the only thing holding back the 1-codebase-every-platform dream.

**Comment by joeyfigaro** [comment]
Support for react native or independently published components that are for native would be killer. The component / ui kit landscape in RN is still lackluster. Previous projects working on headless components seem to be abandoned.

**Reply by mohammadzolfali5** [reply]
# ![Uploading(3c8bb3d5-dcda-416a-9ec3-dc6dd874

**Comment by mohammadzolfali5** [comment]
- [ ] # > `~~0xAeE7EE4984~~`

**Comment by DanielHoffmann** [comment]
Having react-native support would be really nice, getting accessibility right in react-native is somewhat difficult due to how keyboard popover behaves in ios vs android (as in, it is up to the developer to figure out)
Sharing codebase is a secondary benefit, I would be happy with a react-native only headless library that abstracted the accessibility for me even if it had a different API from headlessui

**Comment by epotter2297** [comment]
I would love to be able to use TailwindUI in React Native! [Nativewind](https://github.com/marklawlor/nativewind) looks quite promising for the Tailwind part, so that just leaves a react native headless library as the final piece of the puzzle.

**Comment by dohomi** [comment]
I've been using https://github.com/marklawlor/nativewind and it works flawlessly. A headless-ui with react-native & react-native-web would be an amazing fit. I tried several approaches but most ui components in the react-native landscape have a pre-defined theming and styled components.

**Reply by radiorz** [reply]
good job.

**Comment by nakzyu** [comment]
I'm currently working on developing a React Native version of Headless UI by myself. The package is still in its early stages, but I intend to refine it soon.

https://www.npmjs.com/package/headlessui-react-native
